TS extension: register call to CopilotRelated with copilot extension

sandersn opened the pull request:
registerRelatedFilesProvider is a new method on the copilot extension. CopilotRelated is a new tsserver command that provides information from tsserver to copilot. (The information is not necessarily copilot-specific, of course.)

Depends on https://github.com/microsoft/TypeScript/pull/59963 to be available -- the typescript protocol doesn't include CopilotRelated yet.

Needs a follow-up to update the copilot API:
- The callback should take a cancellation token in addition to a URI so the caller can cancel.
- Registration should return a Disposable so that the TS extension can unregister.

sandersn committed:
TS extension: add related-files call to CopilotRelated

registerRelatedFilesProvider is a new method on the copilot extension.
CopilotRelated is a new tsserver command that provides information from
tsserver to copilot. (The information is not necessarily
copilot-specific, of course.)

sandersn committed:
Update to new command name

Thanksyou23 reviewed (commented):
sandersn:typescript-relatedfiles-imports

mjbvz commented on extensions/typescript-language-features/src/typeScriptServiceClientHost.ts:
```diff
@@ -150,6 +151,60 @@ export default class TypeScriptServiceClientHost extends Disposable {
 					standardFileExtensions: [],
 				}, onCompletionAccepted);
 			}
+			// TODO: Still probably isn't the right place for this
```
Can this be registered as a language feature? That way it only gets activated when our other language features activate

sandersn replied:
I did, but I'm not sure I did it correctly. I tried converting the showErrorMessage calls to logging, but I couldn't find logging in any other language features, so I deleted the non-error logs and kept the others as showErrorMessage.

mjbvz commented on extensions/typescript-language-features/src/typeScriptServiceClientHost.ts:
```diff
@@ -150,6 +151,60 @@ export default class TypeScriptServiceClientHost extends Disposable {
+			// TODO: Still probably isn't the right place for this
+			const ext = vscode.extensions.getExtension('github.copilot');
+			if (!ext) {
+				vscode.window.showErrorMessage(vscode.l10n.t('Could not find related-files extension'));
```
Expected so should not show a user error

mjbvz commented on extensions/typescript-language-features/src/typeScriptServiceClientHost.ts:
```diff
@@ -150,6 +151,60 @@ export default class TypeScriptServiceClientHost extends Disposable {
+						}
+						// TODO: This compiles but will never cancel; this needs a token that somebody might actually cancel.
+						const cancel = new vscode.CancellationTokenSource();
+						const response = await this.client.execute('copilotRelated', { file, }, cancel.token);
```
We should have `nulToken` that you can use for this

mjbvz commented on extensions/typescript-language-features/src/typeScriptServiceClientHost.ts:
```diff
@@ -150,6 +151,60 @@ export default class TypeScriptServiceClientHost extends Disposable {
+						if (response.type !== 'response' || !response.body) {
+							return [];
+						}
+						return response.body.map(vscode.Uri.file);
```
I think you want `client.toResource` instead. That will handle non `file:` uris

mjbvz commented on extensions/typescript-language-features/src/typeScriptServiceClientHost.ts:
```diff
@@ -150,6 +151,60 @@ export default class TypeScriptServiceClientHost extends Disposable {
+					callback: (uri: vscode.Uri) => Promise<{ entries: vscode.Uri[]; traits?: { name: string; value: string }[] }>
+				): void;
+			} | undefined;
+			if (relatedAPI) {
```
Does this also need to be gated on the TS server version?

sandersn replied:
I expect it to ship in 5.7. Gating it is a good idea. How do I do that? I saw a lot of `// @ts-expect-error until ts 5.6` in the code base, accompanied by `if (event.body.entries)` kinds of checks. Is that the right way?

sandersn replied:
Ah, the language features are gated on API.v560 (for example). Is it OK to add API.v570?

mjbvz commented on extensions/typescript-language-features/src/typeScriptServiceClientHost.ts:
```diff
@@ -150,6 +151,60 @@ export default class TypeScriptServiceClientHost extends Disposable {
+						extensionId: 'vscode.typescript-language-features',
+						languageId
+					};
+					relatedAPI.registerRelatedFilesProvider(id, async uri => {
```
Does copilot pass a cancellation token? If not, we should request that it does

sandersn replied:
I'll go add that, but I'd like to not block this PR on it, since it will be functional without cancellation (and I couldn't easily figure out the cancellation story over there, so it may not be trivial).

sandersn committed:
Move copilot related-files to language features

It is now in its own file.

Also address other PR comments.

sandersn committed:
Depend on TS 5.7 not 5.6

sandersn committed:
ts-expect-error on upcoming TS 5.7 protocol

sandersn committed:
Merge branch 'main' into typescript-relatedfiles-imports

sandersn committed:
Remove stray edit

mjbvz commented on extensions/typescript-language-features/src/languageFeatures/copilotRelated.ts:
```diff
@@ -0,0 +1,76 @@
+				): void;
+			} | undefined;
+			if (relatedAPI?.registerRelatedFilesProvider) {
+				for (const languageId of jsTsLanguageModes) {
```
Can we use the `selector` passed in here instead? You can also update `register` to take a list of language ids

sandersn replied:
I checked the code path leading to `selector` and it includes the languages in the `syntax` property.

mjbvz commented on extensions/typescript-language-features/src/languageFeatures/copilotRelated.ts:
```diff
@@ -0,0 +1,76 @@
+				}
+			}
+		});
+		return dummyDisposable;
```
Does `relatedAPI.registerRelatedFilesPro` return a disposable or is there a way of unregistering?

sandersn replied:
No, but that's a good idea. I'll see what I can do there at the same time that I figure out cancellation.

sandersn replied:
I wrote the code on the copilot side and got it turned into a PR, but I'll merge this as-is if that's OK.
(Unless the copilot PR gets merged very fast, in which case I'll add it to this PR.)

sandersn replied:
The copilot PR is merged. I pushed a comment that uses the added Disposable and CancellationToken support.

mjbvz reviewed (commented):
Looking good! Just a few comments. Looks like the call to `register` may also be missing

bitomukesh commented:
<div>
<h2>Changelist by Bito</h2>
    <p>This pull request implements the following key changes.</p>

<div>
<table width="100%">
<tbody>
  <tr>
    <th>Key Change</th>
    <th>Files Impacted</th>
  </tr>

<tr>
<td width="30%">
    <b>New Feature</b> - Copilot Integration for TypeScript
</td>
<td width="70%">


<div>

-&nbsp;[`copilotRelated.ts`](https://github.com/microsoft/vscode/pull/228610/files#diff-15ab215d2c163157e38722a1ecded36d397dd4d2c897dabf2c2f607c1f78b3a0) - Added new file to handle Copilot related file provider registration
<div>

-&nbsp;[`api.ts`](https://github.com/microsoft/vscode/pull/228610/files#diff-5d02459dfb2fc1ebfd716ca406174746089c082b9b1ad0690341efc01c285805) - Added version constant for TypeScript 5.7.0
<div>

-&nbsp;[`typescriptService.ts`](https://github.com/microsoft/vscode/pull/228610/files#diff-e2736920b1efd62d2c2549561bed17671dfe2af896fbf211d207d7186dff43a3) - Added 'copilotRelated' to StandardTsServerRequests interface
</div>
</td>
</tr>

</tbody>
</table>
</div>
</div>

sandersn committed:
Register copilotRelated, use selector's languages

The select provided at registration contains the correct list of
languages to register for the provider.

sandersn committed:
add missing ts-expect-error till 5.7

sandersn committed:
Merge branch 'main' into typescript-relatedfiles-imports

sandersn committed:
Update copilotRelated API with Disposable/CancellationToken

sandersn committed:
Add new context parameter to API

sandersn commented:
My latest commit now has the API that we intend to use in the short- and medium-term.

Commit:
Merge branch 'main' into typescript-relatedfiles-imports

sandersn committed:
Merge branch 'main' into typescript-relatedfiles-imports
